flink-kubernetes/src/test/java/org/apache/flink/kubernetes/highavailability/KubernetesStateHandleStoreTest.java

```diff
@@ -1119,4 +1166,34 @@ private TestingLongStateHandleHelper.LongStateHandle addDeletingEntry(
         configMap.getData().put(key, deleting);
         return state;
     }
+
+    private static CompletableFuture<Boolean> retryWithFirstFailedK8sOperation(
+            Function<KubernetesConfigMap, Optional<KubernetesConfigMap>> function,
+            KubernetesConfigMap leaderConfigMap) {
+        final AtomicInteger callbackInvocationCount = new AtomicInteger(0);
+        final CompletableFuture<Boolean> result =
+                FutureUtils.retry(
+                        () ->
+                                CompletableFuture.supplyAsync(
+                                        () -> {
+                                            callbackInvocationCount.incrementAndGet();
+                                            function.apply(leaderConfigMap);
+                                            if (callbackInvocationCount.get() == 1) {
+                                                throw new KubernetesClientException(
+                                                        "Expected exception to simulate unstable "
+                                                                + "kubernetes client operation");
+                                            }
+                                            return true;
+                                        },
+                                        Executors.newDirectExecutorService()),
+                        KubernetesConfigOptions.KUBERNETES_TRANSACTIONAL_OPERATION_MAX_RETRIES
+                                .defaultValue(),
+                        t ->
+                                ExceptionUtils.findThrowable(t, KubernetesClientException.class)
+                                        .isPresent(),
+                        Executors.newDirectExecutorService());
+        assertThat(callbackInvocationCount.get(), is(2));
+        assertThat(result.isDone(), is(true));
```

Comment:
Same as for the [related PR](https://github.com/apache/flink/pull/20684#discussion_r955768201) in `master`: We might want to check for completion instead of the result being done here.

Reply:
Since we are not using `assertj` in branch 1.15, do you think it is really necessary to also assert the exception here?

```
assertThat(result.isCompletedExceptionally(), is(false));
```

Reply:
interesting. I really would have assumed that there is something like `CompletableFuture::isCompleted()`. :thinking:  I guess, `assertThat(result.get(), is(true));` would be the proper assertion then. WDYT?

Reply:
Given that the exception of the `result` will be eventually thrown in the `addAndLock`, I do not think it is really necessary to check exception here. But it should be harmless.

```
        try {
            assertThat(result.get(), is(true));
        } catch (Exception ex) {
            fail("Exception should not be thrown.");
        }
```

Reply:
The above codes may hide the exception of future. Maybe I need to throw the exception directly in `retryWithFirstFailedK8sOperation`.

Reply:
It seems that I find something.

```
    /**
     * Returns {@code true} if this task completed.
     *
     * Completion may be due to normal termination, an exception, or
     * cancellation -- in all of these cases, this method will return
     * {@code true}.
     *
     * @return {@code true} if this task completed
     */
    boolean isDone();
```

Reply:
Since we expect the `result` is already completed without `.get()`, I prefer to use `assertThat(result.isDone() && !result.isCompletedExceptionally(), is(true));`

Reply:
I see, fair point. But then, you might need to do `assertThat(result.isDone() && !result.isCompletedExceptionally() && !result.isCancelled(), is(true));` to be more precise. You might want to put it into separate asserts instead of having a long boolean condition like that for readability purposes...

Reply:
Make sense. I will update this PR.